[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59062
* __->__ #58935

I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

swolchok committed:
[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors

I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -296,6 +300,46 @@ struct BoxedKernelWrapper<
+// references instead.
+// Note: all signatures matching this pattern are are assumed to be for such ops.
+// This assumption permits the generated BoxedKernelWrapper specializations to simply
+// return out arguments.
```
oh this is so cute, you're using the fact that the return type is a `const Tensor&` to infer that the first argument is what we previously treated as a mutable tensor ref (otherwise it is indistinguishable)

cc @bhosmer

swolchok committed:
Update on "[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors"


I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors"


I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors"


I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors"


I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors"


I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors"


I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors"


I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors"


I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors"


I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors"


I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors"


I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors"


I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors"


I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors"


I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors"


I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors"


I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

[ghstack-poisoned]

swolchok commented:
test_cpp_extensions_aot_no_ninja failure is likely a true positive; DLL load time is when we would find out about a signature mismatch

swolchok committed:
Update on "[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors"


I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors"


I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors"


I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

[ghstack-poisoned]

swolchok commented:
https://github.com/pytorch/xla/issues/3002 opened for the XLA break

swolchok committed:
Update on "[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors"


I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors"


I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors"


I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make arithmetic ops etc. use_const_ref_for_mutable_tensors"


I started off trying to make mv, dot, mm, and bmm const-correct, and the interdependencies between operators meant that I had to fix a bunch more stuff too. I erred on the side of just fixing any one-off thing that came up during this migration.

Differential Revision: [D28687558](https://our.internmc.facebook.com/intern/diff/D28687558/)

[ghstack-poisoned]
